src/System.Net.Security/src/System/Net/Security/SslState.cs

```diff
@@ -1449,6 +1420,33 @@ private void FinishHandshake(Exception e, AsyncProtocolRequest asyncRequest)
             }
         }
 
+        private void ProcessPendingWrite()
+        {
+            object obj = _queuedWriteStateRequest;
+            _queuedWriteStateRequest = null;
+
+            switch (obj)
+            {
+                case null:
+                    // We finished before Write has grabbed the lock.
+                    return;
+                case LazyAsyncResult lazy:
+                    // Sync write is waiting on other thread.
+                    lazy.InvokeCallback();
+                    return;
+                case TaskCompletionSource<int> completionSource:
+                    //Async write is pending, should it be started on another thread?
+                    //Or is it fine to do it here
+                    completionSource.SetResult(0);
+                    return;
+                default:
+                    // Async write is pending, start it on other thread.
+                    // Consider: we could start it in on this thread but that will delay THIS handshake completion
+                    ThreadPool.QueueUserWorkItem(new WaitCallback(CompleteRequestWaitCallback), obj);
```

Comment:
Can we cache this delegate rather than allocating it on each write that pends?

Reply:
We could you mean that last delegate? Its the original code move to pattern matching and is super infrequent. But I will change

Reply:
Right, I meant the WaitCallback.